parallel-20200622/src/env_parallel.1
.\" Automatically generated by Pod::Man 4.09 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "ENV_PARALLEL 1"
.TH ENV_PARALLEL 1 "2020-01-27" "20200122" "parallel"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
env_parallel \- export environment to GNU parallel
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBenv_parallel\fR [\-\-record\-env|\-\-session|\-\-end\-session]
  [options for \s-1GNU\s0 Parallel]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBenv_parallel\fR is a shell function that exports the current
environment to \s-1GNU\s0 \fBparallel\fR.
.PP
If the shell function is not loaded, a dummy script will be run
instead that explains how to install the function.
.PP
\&\fBenv_parallel\fR is 100 ms slower at startup than pure \s-1GNU\s0
\&\fBparallel\fR, and takes up to 30% longer to start a job (typically 15 ms).
.PP
Due to the problem with environment space (see below) the recommended
usage is either:
.PP
.Vb 2
\&  # Do \-\-record\-env into $PARALLEL_IGNORED_NAMES
\&  env_parallel \-\-session
\&
\&  # Define whatever you want to use
\&  alias myalias=echo
\&  myvar=it
\&  myfunc() { myalias $1 $myvar works.; }
\&
\&  # env_parallel will not export names in $PARALLEL_IGNORED_NAMES
\&  env_parallel \-S localhost myfunc ::: Yay,
.Ve
.PP
Or:
.PP
.Vb 2
\&  # Record the "clean" environment (this only needs to be run once)
\&  env_parallel \-\-record\-env
\&
\&  # Optionally edit ~/.parallel/ignored_vars (only needed once)
\&
\&  # Define whatever you want to use
\&  alias myalias=echo
\&  myvar=it
\&  myfunc() { myalias $1 $myvar works.; }
\&
\&  # Use \-\-env _ to only transfer the names not in the "empty" environment
\&  env_parallel \-\-env _ \-S localhost myfunc ::: Yay,
.Ve
.PP
In \fBcsh\fR \fB\-\-session\fR is not supported:
.PP
.Vb 2
\&  # Record the "clean" environment (this only needs to be run once)
\&  env_parallel \-\-record\-env
\&
\&  # Optionally edit ~/.parallel/ignored_vars (only needed once)
\&
\&  # Define whatever you want to use
\&  alias myalias \*(Aqecho \e!* $myvar works.\*(Aq
\&  set myvar=it
\&
\&  # Use \-\-env _ to only transfer the names not in the "empty" environment
\&  env_parallel \-\-env _ \-S localhost myalias ::: Yay,
.Ve
.SS "Environment space"
.IX Subsection "Environment space"
By default \fBenv_parallel\fR will export all environment variables,
arrays, aliases, functions and shell options (see details for the
individual shells below).
.PP
But this only works if the size of the current environment is smaller
than the maximal length of a command and smaller than half of the max
if running remotely. E.g. The max size of Bash's command is 128 \s-1KB,\s0 so
\&\fBenv_parallel\fR will fail if '\fBset | wc \-c\fR' is bigger than 128
\&\s-1KB.\s0 Technically the limit is in \fIexecve\fR\|(1) which IPC::open3 uses.
.PP
Bash completion functions are well-known for taking up well over 128
\&\s-1KB\s0 of environment space and the primary reason for causing
\&\fBenv_parallel\fR to fail.
.PP
Instead you can use \fB\-\-env\fR to specify which variables, arrays,
aliases and functions to export as this will only export those with
the given name. Or follow the recommended usage in shown in
\&\s-1DESCRIPTION.\s0
.SH "OPTIONS"
.IX Header "OPTIONS"
Same as \s-1GNU\s0 \fBparallel\fR in addition to these:
.IP "\fB\-\-end\-session\fR" 4
.IX Item "--end-session"
Undo last \fB\-\-session\fR
.IP "\fB\-\-record\-env\fR" 4
.IX Item "--record-env"
Record all names currently defined to be ignored every time running
\&\fBenv_parallel\fR in the future.
.IP "\fB\-\-session\fR" 4
.IX Item "--session"
Ignore all names currently defined. Aliases, variables, arrays, and
functions currently defined will not be transferred.
.Sp
But names defined \fIafter\fR running \fBparallel \-\-session\fR \fIwill\fR be
transferred.
.Sp
This is only valid in the running shell, and can be undone with
\&\fBparallel \-\-end\-session\fR.
.Sp
You can run multiple \fB\-\-session\fR inside each other:
.Sp
.Vb 10
\&  env_parallel \-\-session
\&  var=not
\&  # var is transferred
\&  env_parallel \-Slocalhost \*(Aqecho var is $var\*(Aq ::: ignored
\&  env_parallel \-\-session
\&  # var is not transferred
\&  env_parallel \-Slocalhost \*(Aqecho var is $var\*(Aq ::: ignored
\&  env_parallel \-\-end\-session
\&  # var is transferred again
\&  env_parallel \-Slocalhost \*(Aqecho var is $var\*(Aq ::: ignored
.Ve
.SH "SUPPORTED SHELLS"
.IX Header "SUPPORTED SHELLS"
.SS "Ash"
.IX Subsection "Ash"
\fIInstallation\fR
.IX Subsection "Installation"
.PP
Put this in \f(CW$HOME\fR/.profile:
.PP
.Vb 1
\&  . \`which env_parallel.ash\`
.Ve
.PP
E.g. by doing:
.PP
.Vb 1
\&  echo \*(Aq. \`which env_parallel.ash\`\*(Aq >> $HOME/.profile
.Ve
.PP
\fISupported use\fR
.IX Subsection "Supported use"
.PP
\&\fB\-\-env\fR is supported to export only the variable, or alias with the
given name. Multiple \fB\-\-env\fRs can be given.
.PP
\&\fB\-\-session\fR is supported.
.IP "aliases" 8
.IX Item "aliases"
.Vb 5
\&  alias myecho=\*(Aqecho aliases\*(Aq
\&  env_parallel myecho ::: work
\&  env_parallel \-S server myecho ::: work
\&  env_parallel \-\-env myecho myecho ::: work
\&  env_parallel \-\-env myecho \-S server myecho ::: work
\&
\&  alias multiline=\*(Aqecho multiline
\&    echo aliases\*(Aq
\&  env_parallel multiline ::: work
\&  env_parallel \-S server multiline ::: work
\&  env_parallel \-\-env multiline multiline ::: work
\&  env_parallel \-\-env multiline \-S server multiline ::: work
.Ve
.IP "functions" 8
.IX Item "functions"
.Vb 1
\&  ash cannot list defined functions \- thus is not supported.
.Ve
.IP "variables" 8
.IX Item "variables"
.Vb 5
\&  myvar=variables
\&  env_parallel echo \*(Aq$myvar\*(Aq ::: work
\&  env_parallel \-S server echo \*(Aq$myvar\*(Aq ::: work
\&  env_parallel \-\-env myvar echo \*(Aq$myvar\*(Aq ::: work
\&  env_parallel \-\-env myvar \-S server echo \*(Aq$myvar\*(Aq ::: work
.Ve
.IP "arrays" 8
.IX Item "arrays"
Arrays are not supported by Ash.
.SS "Bash"
.IX Subsection "Bash"
\fIInstallation\fR

parallel-20200622/NEWS
20200622

New in this release:

* No new functionality

* Bug fixes and man page updates.

News about GNU Parallel:

* GNU Parallel is used at NERSC: https://www.nersc.gov/assets/Uploads/08-Workflows-20200616.pdf

* Resumable batch jobs with GNU Parallel https://igor.io/parallel/

* Lab Meeting technical talk https://www.slideshare.net/hoffmanlab/gnu-parallel-194030490

* Open Sky Software Website Credits https://www.openskysoftware.com/site-credits.htm

* Speed up your Python Scripts using Ubuntu and GNU Parallel https://ecce.esri.ca/wpecce/2019/12/24/speed-up-your-python-scripts-using-ubuntu-and-gnu-parallel/

* Indispensable command-line tools https://negfeedback.blogspot.com/2020/05/indispensable-command-line-tools.html

* GNU Parallel for simulations http://pdebuyl.be/blog/2020/gnu-parallel-for-simulations.html

* Introduction to GNU parallel https://bioinformaticsworkbook.org/Appendix/GNUparallel/GNU_parallel_examples.html


20200522

New in this release:

* While running a job $PARALLEL_JOBSLOT is the jobslot of the job. It
  is equal to {%} unless the job is being retried. See {%} for
  details.

* While running a job $PARALLEL_SSHLOGIN is the sshlogin line with
  number of cores removed. E.g. '4//usr/bin/specialssh user@host'
  becomes: '/usr/bin/specialssh user@host'

* While running a job $PARALLEL_SSHHOST is the host part of an
  sshlogin line. E.g. '4//usr/bin/specialssh user@host' becomes:
  'host'

* --plus activates the replacement strings {slot} = $PARALLEL_JOBSLOT,
  {sshlogin} = $PARALLEL_SSHLOGIN, {host} = $PARALLEL_SSHHOST

* Bug fixes and man page updates.

News about GNU Parallel:

* Portable Batch System (PBS) & GNU Parallel - Running a Program
  Multiple Times in Parallel
  https://www.youtube.com/watch?v=6ccbWu6Befo

* GNU Parallel przykład https://www.youtube.com/watch?v=gs_wG4Kt2G4

* Demo of LINUX APP - GNU PARALLEL - running multiple Gstreamer webcam
  .sh scripts with only 1 command
  https://www.youtube.com/watch?v=trQuA_wmWjg


20200422

This release celebrates GNU Parallel's 10 years as a GNU tool:

> commit ed2dfb1043768154d4e7678e01e10287155fa834
> Author: Ole Tange <ole@tange.dk>
> Date:   Thu Apr 22 01:23:00 2010 +0200
>
>    Name change: Parallel is now GNU Parallel.
>    Basic structure for sshlogin and sshloginfile.

New in this release:

* parsort makes GNU sort run faster for files with more than 1M lines.

* Bug fixes and man page updates.

News about GNU Parallel:

* Several clusters have updated their documentation of GNU Parallel:
  https://www.grid5000.fr/w/GNU_Parallel
  https://research-it.berkeley.edu/virtual-training-running-jobs-parallel-savio
  http://www.hpc.lsu.edu/training/archive/tutorials.php

* add_pho uses GNU Parallel: https://github.com/SeedlingsBabylab/add_pho

* GNU parallel使用笔记https://zhuanlan.zhihu.com/p/37340011

* Virtual Workshop: Running Jobs in Parallel on Savio
  https://ais.berkeley.edu/events/virtual-workshop-running-jobs-parallel-savio/2020-04-20

* Free Concurrency with GNU Parallel
  https://phili.pe/posts/free-concurrency-with-gnu-parallel/


20200322

* Bug fixes and man page updates.

News about GNU Parallel:

* The GNU Parallel course in Copenhagen is cancelled
  https://www.prosa.dk/nc/arrangementer/arrangement/gnu-parallel-med-ole-tange/

* The GNU Parallel course in Århus is cancelled
  https://www.prosa.dk/nc/arrangementer/arrangement/gnu-parallel-og-parallelisering-i-unix-shellen/

* The 10 year anniversary may be cancelled, see
  https://www.gnu.org/software/parallel/10-years-anniversary.html

* Running shell commands in parallel, via GNU Parallel
  https://dev.to/saveriomiroddi/running-shell-commands-in-parallel-via-gnu-parallel-3a72

* Faster nmap scanning with the help of GNU parallel
  https://badshah.io/experiment/faster-nmap-scanning-with-the-help-of-gnu-parallel/


20200222

* Bug fixes and man page updates.


20200122

* --blocktimeout dur - Time out for reading block when using
  --pipe. If it takes longer than dur to read a full block, use the
  partial block read so far.

* Bug fixes and man page updates.

News about GNU Parallel:

* GNU Parallel course in Copenhagen
  https://www.prosa.dk/nc/arrangementer/arrangement/gnu-parallel-med-ole-tange/

* GNU Parallel course in Århus
  https://www.prosa.dk/nc/arrangementer/arrangement/gnu-parallel-og-parallelisering-i-unix-shellen/

* GNU Parallel pour accélérer vos process sous Linux
  https://www.yvonh.com/gnu-parallel-pour-accelerer-vos-process-sous-linux/

* How to copy a file to multiple directories in Linux
  https://net2.com/how-to-copy-a-file-to-multiple-directories-in-linux/

* Running linux commands in parallel https://dev.to/voyeg3r/runing-linux-commands-in-parallel-4ff8


20191222

* GNU Parallel course in Copenhagen
  https://www.prosa.dk/nc/arrangementer/arrangement/gnu-parallel-med-ole-tange/


parallel-20200622/src/env_parallel.sh
#!/usr/bin/env sh

# This file must be sourced in sh:
#
#   . `which env_parallel.sh`
#
# after which 'env_parallel' works
#
#
# Copyright (C) 2016-2020 Ole Tange, http://ole.tange.dk and Free
# Software Foundation, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>
# or write to the Free Software Foundation, Inc., 51 Franklin St,
# Fifth Floor, Boston, MA 02110-1301 USA

env_parallel() {
    # env_parallel.sh

    _names_of_ALIASES() {
	# alias fails on Unixware 5
	for _i in `alias 2>/dev/null | perl -ne 's/^alias //;s/^(\S+)=.*/$1/ && print' 2>/dev/null`; do
	    # Check if this name really is an alias
	    # or just part of a multiline alias definition
	    if alias $_i >/dev/null 2>/dev/null; then
		echo $_i
	    fi
	done
    }
    _bodies_of_ALIASES() {
	# alias may return:
	#   myalias='definition' (GNU/Linux ash)
	#   alias myalias='definition' (FreeBSD ash)
	# so remove 'alias ' from first line
	for _i in "$@"; do
		echo 'alias '"`alias $_i | perl -pe '1..1 and s/^alias //'`"
	done
    }
    _names_of_maybe_FUNCTIONS() {
	set | perl -ne '/^([A-Z_0-9]+)\s*\(\)\s*\{?$/i and print "$1\n"'
    }
    _names_of_FUNCTIONS() {
	# myfunc is a function
	LANG=C type `_names_of_maybe_FUNCTIONS` |
	    perl -ne '/^(\S+) is a function$/ and not $seen{$1}++ and print "$1\n"'
    }
    _bodies_of_FUNCTIONS() {
	LANG=C type "$@" | perl -ne '/^(\S+) is a function$/ or print'
    }
    _names_of_VARIABLES() {
	# This may screw up if variables contain \n and =
	set | perl -ne 's/^(\S+?)=.*/$1/ and print;'
    }
    _bodies_of_VARIABLES() {
	# Crappy typeset -p
	for _i in "$@"
	do
	    perl -e 'print @ARGV' "$_i="
	    eval echo \"\$$_i\" | perl -e '$/=undef; $a=<>; chop($a); print $a' |
		perl -pe 's/[\002-\011\013-\032\\\#\?\`\(\)\{\}\[\]\^\*\<\=\>\~\|\; \"\!\$\&\202-\377]/\\$&/go;'"s/'/\\\'/g; s/[\n]/'\\n'/go;";
	    echo
	done
    }
    _ignore_HARDCODED() {
	# These names cannot be detected
	echo '(_|TIMEOUT)'
    }
    _ignore_READONLY() {
	readonly | perl -e '@r = map {
                chomp;
                # sh on UnixWare: readonly TIMEOUT
	        # ash: readonly var='val'
	        # ksh: var='val'
                s/^(readonly )?([^= ]*)(=.*|)$/$2/ or
	        # bash: declare -ar BASH_VERSINFO=([0]="4" [1]="4")
	        # zsh: typeset -r var='val'
                  s/^\S+\s+\S+\s+(\S[^=]*)(=.*|$)/$1/;
                $_ } <>;
            $vars = join "|",map { quotemeta $_ } @r;
            print $vars ? "($vars)" : "(,,nO,,VaRs,,)";
            '
    }
    _remove_bad_NAMES() {
	# Do not transfer vars and funcs from env_parallel
	_ignore_RO="`_ignore_READONLY`"
	_ignore_HARD="`_ignore_HARDCODED`"
	# Macos-grep does not like long patterns
	# Old Solaris grep does not support -E
	# Perl Version of:
	# grep -Ev '^(...)$' |
	perl -ne '/^(
		     PARALLEL_ENV|
		     PARALLEL_TMP|
		     _alias_NAMES|
		     _bodies_of_ALIASES|
		     _bodies_of_FUNCTIONS|
		     _bodies_of_VARIABLES|
		     _error_PAR|
		     _function_NAMES|
		     _get_ignored_VARS|
		     _grep_REGEXP|
		     _ignore_HARD|
		     _ignore_HARDCODED|
		     _ignore_READONLY|
		     _ignore_RO|
		     _ignore_UNDERSCORE|
		     _list_alias_BODIES|
		     _list_function_BODIES|
		     _list_variable_VALUES|
		     _make_grep_REGEXP|
		     _names_of_ALIASES|
		     _names_of_FUNCTIONS|
		     _names_of_VARIABLES|
		     _names_of_maybe_FUNCTIONS|
		     _parallel_exit_CODE|
		     _prefix_PARALLEL_ENV|
		     _prefix_PARALLEL_ENV|
		     _remove_bad_NAMES|
		     _remove_readonly|
		     _variable_NAMES|
		     _warning_PAR|
		     _which_PAR)$/x and next;
	    # Filter names matching --env
	    /^'"$_grep_REGEXP"'$/ or next;
            /^'"$_ignore_UNDERSCORE"'$/ and next;
	    # Remove readonly variables
            /^'"$_ignore_RO"'$/ and next;
            /^'"$_ignore_HARD"'$/ and next;
            print;'
    }
    _get_ignored_VARS() {
        perl -e '
            for(@ARGV){
                $next_is_env and push @envvar, split/,/, $_;
                $next_is_env=/^--env$/;
            }
            if(grep { /^_$/ } @envvar) {
                if(not open(IN, "<", "$ENV{HOME}/.parallel/ignored_vars")) {
             	    print STDERR "parallel: Error: ",
            	    "Run \"parallel --record-env\" in a clean environment first.\n";
                } else {
            	    chomp(@ignored_vars = <IN>);
                }
            }
            if($ENV{PARALLEL_IGNORED_NAMES}) {
                push @ignored_vars, split/\s+/, $ENV{PARALLEL_IGNORED_NAMES};
                chomp @ignored_vars;
            }
            $vars = join "|",map { quotemeta $_ } @ignored_vars;
	    print $vars ? "($vars)" : "(,,nO,,VaRs,,)";
            ' -- "$@"
    }

    # Get the --env variables if set
    # --env _ should be ignored
    # and convert  a b c  to (a|b|c)
    # If --env not set: Match everything (.*)
    _make_grep_REGEXP() {
        perl -e '
            for(@ARGV){
                /^_$/ and $next_is_env = 0;
                $next_is_env and push @envvar, split/,/, $_;
                $next_is_env = /^--env$/;
            }
            $vars = join "|",map { quotemeta $_ } @envvar;
            print $vars ? "($vars)" : "(.*)";
            ' -- "$@"
    }
    _which_PAR() {
	# type returns:
	#   ll is an alias for ls -l (in ash)
	#   bash is a tracked alias for /bin/bash
	#   true is a shell builtin (in bash)
	#   myfunc is a function (in bash)
	#   myfunc is a shell function (in zsh)
	#   which is /usr/bin/which (in sh, bash)
	#   which is hashed (/usr/bin/which)
	#   gi is aliased to `grep -i' (in bash)
	#   aliased to `alias | /usr/bin/which --tty-only --read-alias --show-dot --show-tilde'
	# Return 0 if found, 1 otherwise
	LANG=C type "$@" |
	    perl -pe '$exit += (s/ is an alias for .*// ||
	                        s/ is aliased to .*// ||
                                s/ is a function// ||
                                s/ is a shell function// ||
                                s/ is a shell builtin// ||
                                s/.* is hashed .(\S+).$/$1/ ||
                                s/.* is (a tracked alias for )?//);
                      END { exit not $exit }'
    }
    _warning_PAR() {
	echo "env_parallel: Warning: $@" >&2
    }
    _error_PAR() {
	echo "env_parallel: Error: $@" >&2
    }

    if _which_PAR parallel >/dev/null; then
	true parallel found in path
    else
	_error_PAR 'parallel must be in $PATH.'
	return 255
    fi

    # Grep regexp for vars given by --env
    _grep_REGEXP="`_make_grep_REGEXP \"$@\"`"
    unset _make_grep_REGEXP
    
    # Deal with --env _
    _ignore_UNDERSCORE="`_get_ignored_VARS \"$@\"`"
    unset _get_ignored_VARS

    # --record-env
    if perl -e 'exit grep { /^--record-env$/ } @ARGV' -- "$@"; then
	true skip
    else
	(_names_of_ALIASES;
	 _names_of_FUNCTIONS;
	 _names_of_VARIABLES) |
	    cat > $HOME/.parallel/ignored_vars
	return 0
    fi

    # --session
    if perl -e 'exit grep { /^--session$/ } @ARGV' -- "$@"; then
	true skip
    else
	# Insert ::: between each level of session
	# so you can pop off the last ::: at --end-session
	PARALLEL_IGNORED_NAMES="`echo \"$PARALLEL_IGNORED_NAMES\";
          echo :::;
          (_names_of_ALIASES;
	   _names_of_FUNCTIONS;
	   _names_of_VARIABLES) | perl -ne '
	    BEGIN{
	      map { $ignored_vars{$_}++ }
                split/\s+/, $ENV{PARALLEL_IGNORED_NAMES};
	    }
	    chomp;
	    for(split/\s+/) {
	      if(not $ignored_vars{$_}) {
	        print $_,\"\\n\";
	      }
            }
	    '`"
	export PARALLEL_IGNORED_NAMES
	return 0
    fi
    if perl -e 'exit grep { /^--end.?session$/ } @ARGV' -- "$@"; then
	true skip
    else
	# Pop off last ::: from PARALLEL_IGNORED_NAMES
	PARALLEL_IGNORED_NAMES="`perl -e '
          $ENV{PARALLEL_IGNORED_NAMES} =~ s/(.*):::.*?$/$1/s;
	  print $ENV{PARALLEL_IGNORED_NAMES}
        '`"
	return 0
    fi
    # Grep alias names
    _alias_NAMES="`_names_of_ALIASES | _remove_bad_NAMES | xargs echo`"
    _list_alias_BODIES="_bodies_of_ALIASES $_alias_NAMES"
    if [ "$_alias_NAMES" = "" ] ; then
	# no aliases selected
	_list_alias_BODIES="true"
    fi
    unset _alias_NAMES

    # Grep function names
    _function_NAMES="`_names_of_FUNCTIONS | _remove_bad_NAMES | xargs echo`"
    _list_function_BODIES="_bodies_of_FUNCTIONS $_function_NAMES"
    if [ "$_function_NAMES" = "" ] ; then
	# no functions selected
	_list_function_BODIES="true"
    fi
    unset _function_NAMES

    # Grep variable names
    _variable_NAMES="`_names_of_VARIABLES | _remove_bad_NAMES | xargs echo`"
    _list_variable_VALUES="_bodies_of_VARIABLES $_variable_NAMES"
    if [ "$_variable_NAMES" = "" ] ; then
	# no variables selected
	_list_variable_VALUES="true"
    fi
    unset _variable_NAMES

    PARALLEL_ENV="`
        $_list_alias_BODIES;
        $_list_function_BODIES;
        $_list_variable_VALUES;
    `"
    export PARALLEL_ENV
    unset _list_alias_BODIES _list_variable_VALUES _list_function_BODIES
    unset _bodies_of_ALIASES _bodies_of_VARIABLES _bodies_of_FUNCTIONS
    unset _names_of_ALIASES _names_of_VARIABLES _names_of_FUNCTIONS
    unset _ignore_HARDCODED _ignore_READONLY _ignore_UNDERSCORE
    unset _remove_bad_NAMES _grep_REGEXP
    unset _prefix_PARALLEL_ENV
    # Test if environment is too big
    if `_which_PAR true` >/dev/null 2>/dev/null ; then
	parallel "$@"
	_parallel_exit_CODE=$?
	# Clean up variables/functions
	unset PARALLEL_ENV
	unset _which_PAR _which_TRUE
	unset _warning_PAR _error_PAR
	# Unset _parallel_exit_CODE before return
	eval "unset _parallel_exit_CODE; return $_parallel_exit_CODE"
    else
	unset PARALLEL_ENV;
	_error_PAR "Your environment is too big."
	_error_PAR "You can try 3 different approaches:"
	_error_PAR "1. Run 'env_parallel --session' before you set"
	_error_PAR "   variables or define functions."
	_error_PAR "2. Use --env and only mention the names to copy."
	_error_PAR "3. Try running this in a clean environment once:"
	_error_PAR "     env_parallel --record-env"
	_error_PAR "   And then use '--env _'"
	_error_PAR "For details see: man env_parallel"
	return 255
    fi
}

parset() {
    _parset_PARALLEL_PRG=parallel
    _parset_main "$@"
}

env_parset() {
    _parset_PARALLEL_PRG=env_parallel
    _parset_main "$@"
}

_parset_main() {
    # If $1 contains ',' or space:
    #   Split on , to get the destination variable names
    # If $1 is a single destination variable name:
    #   Treat it as the name of an array
    #
    #   # Create array named myvar
    #   parset myvar echo ::: {1..10}
    #   echo ${myvar[5]}
    #
    #   # Put output into $var_a $var_b $var_c
    #   varnames=(var_a var_b var_c)
    #   parset "${varnames[*]}" echo ::: {1..3}
    #   echo $var_c
    #
    #   # Put output into $var_a4 $var_b4 $var_c4
    #   parset "var_a4 var_b4 var_c4" echo ::: {1..3}
    #   echo $var_c4

    _make_TEMP() {
	# mktemp does not exist on some OS
	perl -e 'use File::Temp qw(tempfile);
                 $ENV{"TMPDIR"} ||= "/tmp";
                 print((tempfile(DIR=>$ENV{"TMPDIR"}, TEMPLATE => "parXXXXX"))[1])'
    }

    _parset_NAME="$1"
    if [ "$_parset_NAME" = "" ] ; then
	echo parset: Error: No destination variable given. >&2
	echo parset: Error: Try: >&2
	echo parset: Error: ' ' parset myarray echo ::: foo bar >&2
	return 255
    fi
    shift
    echo "$_parset_NAME" |
	perl -ne 'chomp;for (split /[, ]/) {
            # Allow: var_32 var[3]
	    if(not /^[a-zA-Z_][a-zA-Z_0-9]*(\[\d+\])?$/) {
                print STDERR "parset: Error: $_ is an invalid variable name.\n";
                print STDERR "parset: Error: Variable names must be letter followed by letters or digits.\n";
                $exitval = 255;
            }
        }
        exit $exitval;
        ' || return 255
    _exit_FILE=`_make_TEMP`
    if perl -e 'exit not grep /,| /, @ARGV' "$_parset_NAME" ; then
	# $_parset_NAME contains , or space
	# Split on , or space to get the names
	eval "`
	    # Compute results into files
	    ($_parset_PARALLEL_PRG --files -k "$@"; echo $? > "$_exit_FILE") |
		# var1= cat tmpfile1; rm tmpfile1
		# var2= cat tmpfile2; rm tmpfile2
		parallel -q echo {2}='\`cat {1}; rm {1}\`' :::: - :::+ \`
		    echo "$_parset_NAME" |
			perl -pe 's/,/ /g'
			 \`
	    `"
    else
	# $_parset_NAME does not contain , or space
	# => $_parset_NAME is the name of the array to put data into
	# Supported in: bash zsh ksh mksh
	# Arrays do not work in: sh ash dash
	eval "$_parset_NAME=( $(
	    # Compute results into files. Save exit value
	    ($_parset_PARALLEL_PRG --files -k "$@"; echo $? > "$_exit_FILE") |
                perl -pe 'chop;$_="\"\`cat $_; rm $_\`\" "'
            ) )"
    fi
    unset _parset_NAME _parset_PARALLEL_PRG _parallel_exit_CODE
    # Unset _exit_FILE before return
    eval "unset _exit_FILE; return \`cat $_exit_FILE; rm $_exit_FILE\`"
}
